msbuild/Xamarin.iOS.Tasks.Core/Xamarin.iOS.Common.targets

```diff
@@ -376,34 +376,55 @@ Copyright (C) 2013-2016 Xamarin. All rights reserved.
 			_CodesignVerify;
 		</_CoreCodesignDependsOn>
 
-		<CodesignDependsOn>
+		<_BeforeCodeSignForBuildDependsOn>
+			_CreateAppBundle;
+			BeforeCodeSign;
+			CoreCodeSign;
+			AfterCodeSign
+		</_BeforeCodeSignForBuildDependsOn>
+
+		<_BeforeCodeSignForDistributionDependsOn>
 			BeforeCodeSign;
 			CoreCodeSign;
 			AfterCodeSign;
-		</CodesignDependsOn>
+		</_BeforeCodeSignForDistributionDependsOn>
 	</PropertyGroup>
 
 	<Target Name="BeforeCodesign" />
 	<Target Name="CoreCodesign" DependsOnTargets="$(_CoreCodesignDependsOn)" />
 	<Target Name="AfterCodesign" />
 
-	<Target Name="Codesign" Condition="'$(_CanOutputAppBundle)' == 'true'" DependsOnTargets="_CreateAppBundle;$(CodesignDependsOn)" />
+	<Target Name="_BeforeCodeSignForBuild" Condition="'$(IsAppDistribution)' != 'True'" BeforeTargets="Codesign" DependsOnTargets="$(_BeforeCodeSignForBuildDependsOn)" />
+	<Target Name="_BeforeCodeSignForDistribution" Condition="'$(IsAppDistribution)' == 'True'" BeforeTargets="Codesign" DependsOnTargets="$(_BeforeCodeSignForDistributionDependsOn)" />
+	
+	<Target Name="Codesign" Condition="'$(_CanOutputAppBundle)' == 'true'" />
 
 	<PropertyGroup>
-		<CreateIpaDependsOn>
+		<_BeforeCreateIpaDependsOn>
 			_CoreCreateIpa;
 			_PackageOnDemandResources;
 			_ZipIpa
-		</CreateIpaDependsOn>
+		</_BeforeCreateIpaDependsOn>
+		
+		<_BeforeCreateIpaForBuildDependsOn>
+			$(_BeforeCreateIpaDependsOn)
+		</_BeforeCreateIpaForBuildDependsOn>
 
+		<_BeforeCreateIpaForDistributionDependsOn>
+			$(_BeforeCreateIpaDependsOn)
+		</_BeforeCreateIpaForDistributionDependsOn>
```

Comment:
It seems both `_BeforeCreateIpaForBuildDependsOn` and `_BeforeCreateIpaForDistributionDependsOn` are identical, so doesn't this distinction just introduce unnecessary complexity? Or are there plans to make these different in the future?

Reply:
That's a really good point. The main purpose for those is to have specific extensibility points for both actions (create an IPA for distribution or as part of a full build). For instance, from a VS perspective, both are different situations, when creating an IPA for distribution (from an archive) we need to copy the xcarchive file to the Mac to then create the IPA from there, but when doing the same thing as part of a full build there's no archive involved since we'll use the app bundle generated by the build process.

I can add some comments on those property groups to clarify this if needed.

Reply:
Yes, a comment or two would be nice, there are other changes needed as well because quite a few tests are failing.